Fixes #29257 - Remove x-editable-rails

xprazak2 opened the pull request:
Moves the setting inline edit to React using Foreman Modal and Formik form.
This allows us to drop x-editable-rails gem and get rid of one more jquery dependency.
Settings loading is done by helper component, that should be temporary
Loading should be done through API once we move the whole Setting page to React.

Commit:
Fixes #29257 - Remove x-editable-rails

ezr-ondrej commented on app/views/settings/index.html.erb:
```diff
@@ -4,6 +4,9 @@
 first_category = ordered_settings.values.flatten.first.try(:category)
 %>
+
+<%= react_component('SettingRecords', { :settings => grouped_settings(@settings) }) %>
```
Hmm this component has no UI at all right? It's here just to load the settings.
Would it make sense for it to be part of the setting table or if not (I can see why it's easier to do it for all settings at once) should we introduce a general component `SaveMyJsonBlobIntoRedux`?

ezr-ondrej replied:
Oh, this would probably get droped after we will manage to get this to a React Page, right?

xprazak2 replied:
Exactly, I do not think we want to encourage this as a pattern.

ezr-ondrej commented on app/views/settings/index.html.erb:
```diff
@@ -15,28 +18,11 @@ first_category = ordered_settings.values.flatten.first.try(:category)
   <% end %>
 </div>
+
+<%= react_component('SettingUpdateModal') %>
```
I :heart: this approach compared to having form component for each setting :+1:

ezr-ondrej commented on webpack/assets/javascripts/react_app/components/SettingUpdateModal/components/SettingValueField.js:
```diff
@@ -0,0 +1,59 @@
+import { arraySelection } from '../../SettingsTable/SettingsTableHelpers';
+import { renderOptions } from '../../common/forms/SelectHelpers';
+
+const SettingValueField = ({ setting, form, field }) => {
```
Still not ideal, but I believe good enough (all the convoluted statements are hidden in this component, so would be easier to refactor in the future)

ezr-ondrej commented on webpack/assets/javascripts/react_app/components/SettingsTable/SettingsTableReducer.js:
```diff
@@ -0,0 +1,18 @@
+import { SET_UPDATE_SETTING } from './SettingsTableConstants';
+
+const initialState = Immutable({
+  toEdit: {},
```
weird name and weird placement. this is in `SettingTableReducers` right? but it's shared with the update form.
Ideal state would be, in the shared data of all settings table just noting `'editing': <id>` of the setting that should get edited. The Wrapped form would check on that state attribute, and if the id is set, it would get the proper setting from the store.

xprazak2 replied:
I moved the state, but I kept storing the whole setting - I already have access to the setting object and I know I will need it later, that will save me an additional filtering of settings from store.

ezr-ondrej commented on webpack/assets/javascripts/react_app/components/SettingsTable/index.js:
```diff
@@ -0,0 +1,39 @@
+  const { setModalOpen } = useSettingModal();
+
+  const onEditClick = async setting => {
+    await dispatch(setSettingToUpdate(setting));
+    setModalOpen();
```
we could just set the `editing: <id>` by some action, and the modal would pick up on that, so we won't need to call `setModalOpen()` explicitly.

xprazak2 replied:
We use hooks to open or close modals, they have their own slice of redux store:

![redux-modal](https://user-images.githubusercontent.com/2664090/88369828-23a21c00-cd91-11ea-86e0-12c48c3f3b5c.png)

ForemanModal does not currently support the behaviour that you describe.

ezr-ondrej commented on webpack/assets/javascripts/react_app/components/common/forms/InputField.js:
```diff
@@ -0,0 +1,11 @@
+import { FormControl } from 'patternfly-react';
+import PropTypes from 'prop-types';
+
+const InputField = ({ field }) => <FormControl {...field} type="text" />;
```
Do you use this component?

ezr-ondrej commented on webpack/assets/javascripts/react_app/components/common/forms/Select.js:
```diff
@@ -9,6 +8,7 @@ import { noop } from '../../../common/helpers';
 import CommonForm from './CommonForm';
 import { STATUS } from '../../../constants';
 import MessageBox from '../MessageBox';
+import { renderOptions } from './SelectHelpers';
```
Nice refactor :+1:
Should we maybe extract it to a separate PR?
You could even use the same ticket as we've used for #7700 that introduced those methods.

ezr-ondrej reviewed (commented):
In general, I love this, only the modal opening is bit convoluted, but I belive that can be fixed easily.

ezr-ondrej commented:
Also, this https://github.com/theforeman/foreman/blob/bed53cf2cb7f0e10927ab913774457f577ee28c1/app/assets/config/manifest.js#L7 needs to get erased.

Commit:
Refs #29257 - Move state for edited setting

xprazak2 commented:
I extracted https://github.com/theforeman/foreman/pull/7851, I'll rebase on it once it is merged.

ezr-ondrej commented:
I think the `SettingValueSelection` test failure is related, apart of it I think it's good step forward and I would like to get it in `2.2`.
Do you think we will be able to finish this and switch to the Setting react page next week?

xprazak2 commented:
Updated with a test fix.

ezr-ondrej commented:
Should we send some notice to katello, that their test is going to fail after merge?

xprazak2 commented:
Done: https://github.com/Katello/katello/pull/8857

ezr-ondrej commented on app/helpers/settings_helper.rb:
```diff
@@ -40,4 +40,32 @@ def translate_full_name(setting)
+      :full_name => setting.full_name,
+      :config_file => setting.class.config_file,
+      :select_values => setting.select_collection,
+      :value => setting.value,
```
This is using the unencrypted value, so even the encrypted settings are readable now.

ezr-ondrej reviewed (changes requested):
Overall looks great :+1:
I've found one issue though. We are exposing the encrypted settings now :(

sharvit commented on webpack/assets/javascripts/react_app/components/SettingRecords/SettingRecordsSelectors.js:
```diff
@@ -0,0 +1,10 @@
+export const selectSettingsByCategory = (state, category) =>
+  selectSettings(state)[category].map(propsToCamelCase);
+export const selectSettingById = (state, id, category) =>
+  selectSettingsByCategory(state, category).find(setting => setting.id === id);
```
Where using `map/find` in a selector it usually makes sense to memorize them to avoid unnecessary calculations.

Commit:
Refs #29257 - Add memoized selectors

xprazak2 commented:
I added memoized selectors and encrypted settings are now hidden.

ezr-ondrej reviewed (commented):
Wouldn't it be easier just to return encrypted value from the server?
The values are still accessible for (highly) experienced users through the javascript console, right?

xprazak2 commented:
You are right, there would still be ways to get to the values. API just returns a [placeholder](https://github.com/theforeman/foreman/blob/develop/app/models/concerns/hidden_value.rb#L4) for encrypted settings, which complicates updating an existing value. Let me see if I can find a way to deal with this.

ezr-ondrej commented:
> Let me see if I can find a way to deal with this.

Currently we can't edit the value, so I think if we would manage opening the form with empty value on every edit, I think it would be sufficient.
Lets discuss this on the next maintainers meeting?

Commit:
Refs #29257 - Hide encrypted setting values

xprazak2 commented:
I agree showing an empty field is probably the best solution here:


![encrypted-field](https://user-images.githubusercontent.com/2664090/89288127-6dbeb380-d655-11ea-817d-32611cbe2841.png)

ezr-ondrej reviewed (commented):
We need a rebase (I can still see the helper changes).
And we now show the value in the table for collection setting (instead of the label).
But I can see it's not easy to change, so I don't insist, just if you'd have an idea how to do it easily.

ezr-ondrej commented:
The failures are related, but I guess easy to fix (remove the test).

ezr-ondrej commented on webpack/assets/javascripts/react_app/components/SettingRecords/__tests__/SettingRecords.fixtures.js:
```diff
@@ -0,0 +1,212 @@
+    name: 'root_pass',
+    fullName: 'Root password',
+    selectValues: null,
+    value: 'changeme',
```
nit
```suggestion
    value: '****',
```

ezr-ondrej reviewed (changes requested):
~~For some reason it doesn't preselect the setting value in select form~~ :thinking: 
EDIT: ignore me, that was my setup.

ezr-ondrej commented on webpack/assets/javascripts/react_app/components/SettingsTable/SettingsTableHelpers.js:
```diff
@@ -0,0 +1,154 @@
+const formatArraySelectionValue = setting =>
+  formatArraySelection('value', setting);
+
+const formatArraySelection = (attr, setting) => {
```
I'm missing a simmilar formatter for non-grouped options :thinking:
(eg `default_locale`, `default_timezone`)

ezr-ondrej commented on webpack/assets/javascripts/react_app/components/SettingsTable/SettingsTableHelpers.js:
```diff
@@ -0,0 +1,154 @@
+  if (!Array.isArray(selectValues)) {
+    return null;
+  }
+  return deepPropsToCamelCase(selectValues);
```
I don't really get this :thinking:

xprazak2 replied:
It transforms keys of the select options to proper case:

```js
console.log(selectValues)

[
  { value: null, label: 'Select an owner' },
  {
    group_label: 'Users',
    children: [
      { label: 'Admin', value: '4-Users' }
    ],
  },
 {
    group_label: 'Usergroups',
    children: [
      { label: 'Default Usergroup', value: '5-Usergroups' }
    ],
  }
]
```

ezr-ondrej replied:
Ah! I see `group_label => groupLabel` ^_^

ezr-ondrej commented on app/helpers/settings_helper.rb:
```diff
@@ -40,4 +40,33 @@ def translate_full_name(setting)
+      :settings_type => setting.settings_type,
+      :default => setting.default,
+      :readonly => setting.readonly?,
+      :full_name => setting.full_name,
```
The `full_name` is not being translated anywhere else IIUC.
```suggestion
      :full_name => setting.full_name.nil? ? setting.name : _(setting.full_name),
```

ezr-ondrej commented on webpack/assets/javascripts/react_app/components/SettingsTable/components/SettingName.js:
```diff
@@ -0,0 +1,12 @@
+import PropTypes from 'prop-types';
+
+const SettingName = ({ setting }) => (
+  <React.Fragment>{setting.fullName}</React.Fragment>
```
Alternatively to ^^ we could
```suggestion
  <React.Fragment>{setting.fullName ? __(setting.fullName) : setting.name}</React.Fragment>
```

ezr-ondrej commented:
While testing more thoroughly, it seems the `SettingsController#update` differs from `API::SettingsController#update`.
I've noticed it's not possible to nilify setting through the api. As well I'd guess it'll be impossible to set numeric values to string settings (should be just edge cases, but it can be needed).

Commit:
Refs #29257 - Fix minor issues

xprazak2 commented:
I fixed formatting for hash selection and added translation for full name.

I am not sure which settings you tried to set empty values for, I was able to set and remove text for the `login_text` setting.

Numeric values for string settings are out of scope for these changes.

ezr-ondrej commented:
Please just drop the tests for the `SettingsController#update` and I think we are good to go :+1:

Commit:
Refs #29257 - Drop tests

xprazak2 commented:
Updated.

ezr-ondrej reviewed (approved):
Thanks @xprazak2 awesome improvement! :+1: 

Thanks @sharvit for reviewing and @ares for noticing it's definitely demo worthy :)

ezr-ondrej commented:
@tbrisker I've created a packaging PR, just as a reminder https://github.com/theforeman/foreman-packaging/pull/5842
But I believe that droping is automated on release? :thinking:

tbrisker commented:
Since this is touching the UI I'd be happy to wait for integration tests to work again before merging, just to make sure that nothing broke. Also i think the katello test failure might be related?

ezr-ondrej commented:
> Since this is touching the UI I'd be happy to wait for integration tests to work again before merging, just to make sure that nothing broke. Also i think the katello test failure might be related?

@tbrisker well depends on how fast we can fix it and currently it doesn't depend on us, the tests were passing before the last commit, so I don't think we need to block it.
There is a PR to Katello for fixing the test as commented in https://github.com/theforeman/foreman/pull/7775#issuecomment-665720181

tbrisker reviewed (approved):
Packaging-wise this is good to go, it's just dropping a dependency.
